### ISSUE TITLE
Error While building BioGearsEngine/Core from source

### ISSUE THREAD

**Issue by alekhgupta1441** [BioGearsEngine/core#14]
After running command: `cmake --build . --config Release --target scenario_driver`
                                   or  `ninja scenario_driver`
I am getting error while building `Diffusion.h` as:

>[327/345] Building CXX object projects...ars/src/engine/Systems/Diffusion.cpp.o
FAILED: projects/biogears/CMakeFiles/libbiogears.dir/libBiogears/src/engine/Systems/Diffusion.cpp.o 
/usr/bin/g++-7  -DBIOGEARS_COMMON_BUILD_STATIC -DBIOGEARS_THROW_NAN_EXCEPTIONS -DBIOGEARS_THROW_READONLY_EXCEPTIONS -Dbiogears_EXPORTS -I../projects/biogears/libBiogears/include -I../projects/biogears/libBiogears/src -Iprojects/biogears -Iprojects/biogears-common -I../projects/biogears-common/include -I../projects/biogears/libCDM/include -isystem /usr/include/eigen3 -O3 -DNDEBUG -fPIC   -std=gnu++14 -MD -MT projects/biogears/CMakeFiles/libbiogears.dir/libBiogears/src/engine/Systems/Diffusion.cpp.o -MF projects/biogears/CMakeFiles/libbiogears.dir/libBiogears/src/engine/Systems/Diffusion.cpp.o.d -o projects/biogears/CMakeFiles/libbiogears.dir/libBiogears/src/engine/Systems/Diffusion.cpp.o -c ../projects/biogears/libBiogears/src/engine/Systems/Diffusion.cpp
In file included from ../projects/biogears/libBiogears/src/engine/Systems/Diffusion.cpp:12:0:
../projects/biogears/libBiogears/include/biogears/engine/Systems/Diffusion.h:32:47: error: ‘unique_ptr’ in namespace ‘std’ does not name a template type
   static auto make_unique(BioGears& bg)->std::unique_ptr<DiffusionCalculator>;
                                               ^~~~~~~~~~
../projects/biogears/libBiogears/include/biogears/engine/Systems/Diffusion.h:32:47: error: expected ‘;’ at end of member declaration
../projects/biogears/libBiogears/include/biogears/engine/Systems/Diffusion.h:32:57: error: expected unqualified-id before ‘<’ token
   static auto make_unique(BioGears& bg)->std::unique_ptr<DiffusionCalculator>;
                                                         ^
../projects/biogears/libBiogears/src/engine/Systems/Diffusion.cpp:21:61: error: no ‘std::unique_ptr<biogears::DiffusionCalculator> biogears::DiffusionCalculator::make_unique(biogears::BioGears&)’ member function declared in class ‘biogears::DiffusionCalculator’
 auto DiffusionCalculator::make_unique(BioGears& bg) -> std::unique_ptr<DiffusionCalculator>
                                                             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
[332/345] Building CXX object projects...ogears/src/engine/Systems/Energy.cpp.o
>ninja: build stopped: subcommand failed.
 
Generally this occurs due to wrong compiler version for CMAKE_CXX_COMPILER or its PATH.But here std=gnu++14 clearly shown in error log.I have changed even the compiler to g++ as shown in above log,by editing cmake generated file `CMakeCache.txt`.
**Does it require some changes in CMakeLists.txt. as `unique_ptr` was given this feature in higher c++11 versions and by default cmake is taking CXX_COMPILER as gcc only.**

**Comment by alekhgupta1441** [comment]
I am also interested in contributing to BioGears through GSoC this year.I am interested in project "Adding Drug:Atropine and validate their Pharmokinetics curves and their Pharmodynamic Responses".I just want to clarify whether I should be submitting a proposal or not.
I have contacted through mail earlier but got no response.

**Comment by StevenAWhite** [comment]
Please post your compiler version and operating system for any build support ticket.  You could also attach your full CMakeCache.text to assist the support team in identifying your bug.

**Comment by StevenAWhite** [comment]
### TL;DR 
Try adding `#include <memory>` to the top of that file and see if it fixes your issue

### Thoughts

Now, the line referenced

`static auto make_unique(BioGears& bg) -> std::unique_ptr<DiffusionCalculator>;`

is valid C++11 and has no requirements beyond that.  It is not using the std::make_unique call which was added to the standard in C++14, but instead creating a similarly named function which isn't a template in the biogears namespace.  
The error called out is that the type referenced in the return type is not expected to be a template hints it declares the character '<' to be unexpected.  Given the error line truncates the line down to the type name unique_ptr I believe you should read this as confirmation that in your build infrastructure the type unique_ptr is not known when that line is being passed. 

```../projects/biogears/libBiogears/src/engine/Systems/Diffusion.cpp:21:61: error: no ‘std::unique_ptrbiogears::DiffusionCalculator biogears::DiffusionCalculator::make_unique(biogears::BioGears&)’ member function declared in class ‘biogears::DiffusionCalculator’
auto DiffusionCalculator::make_unique(BioGears& bg) -> std::unique_ptr
^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

So, I would conclude your version of the STL does not implicitly include `<memory>` with in the transitive header chain and all of our test compilations systems do.  Now, I find transitive includes bad practice for this reason, so If this resolves your issue I will patch master to explicitly add the STL header.

**Comment by alekhgupta1441** [comment]
Thanks @StevenAWhite  
Including: 
> #inlcude  "memory" 
    
In header file diffusion.h solves the problem.

But there was one more issue while building it from source that cmake version requirement should be upgraded to latest cmake 3.16.0 rather than 3.12.0 as following issue arrised while building using cmake 3.12.0 and resolved after installing latest version of **cmake**.
![Screenshot from 2020-03-18 21-14-06](https://user-images.githubusercontent.com/43706342/77206572-45f30700-6b1d-11ea-99a9-86e0f8a8e632.png)

**Comment by StevenAWhite** [comment]
What documentation did you use for the build.  I've been trying to update
all the cmake refs.  I'm assuming I missed a few

**Comment by alekhgupta1441** [comment]
https://github.com/BioGearsEngine/core/wiki/Using-Ubuntu-package-system
This one.
@StevenAWhite  
Can you please show some light on my above query regarding GSoC.Whether I should give a proposal or not? I would love to contribute to BioGears. I am quite proficient in C++ and have quite some experience with development too.
I have mailed to @ajbaird  but got no response. I just want to clarify.

**Comment by StevenAWhite** [comment]
@alekhgupta1441 

You should submit a proposal based on one of our suggested projects.  If you have a unique project which is not suggested you are welcome to propose that as well.  Include a resume and any relevant project examples you may have out on the net or github.  You can email Austin or MySelf.  We are spinning up a public slack for the GoSC  you can get an invite on our dev domain.

https://www.biogears.dev/
